### ISSUE TITLE
Investigate JS bundle splitting

### ISSUE THREAD

**Issue by muffinresearch** [mozilla/addons-frontend#3734]
There's some possible avenues to speed up the JS by splitting our code into smaller bundles. 

There's two distinct places to look at bundle splitting.

* Split out the vendor code into a separate bundle. We only get a gain from this if we don't change the vendor libs often - given the pace things move at we should consider if we would see much benefit from this. However it might make more sense in combination with the next item.
* Make multiple entry points in the app - this would mean that a specific page would load only the relevant JS for itself rather than loading the JS for the whole site. 

A new issue should be filed for each of these so they can be looked at independently.

This area could have some trade-offs so we should make sure that if we try them, we do each one in a way that allows us to measure the effects of implementing them and we should investigate what others have found in this area before we start.

_Update: Additional considerations_

Another consideration here is to think about the different types of users. On a desktop with a fast connection the big bundle up front might be better, but for a mobile client it could be a completely different story.

We should investigate traffic and usage patterns to see if this is something that we'd be likely to benefit from.

---

**Update for QA:** the about and review guides pages are loaded differently now:

- https://addons-dev.allizom.org/en-US/about
- https://addons-dev.allizom.org/en-US/review_guide

* There should be no visual regression and no issues when navigating to/from these pages.
* There should be NO page reload when loading one of these pages and then clicking another (internal) link in the menu (or the FF logo).
* You _might_ notice a short delay on FF when clicking the "about" link from the homepage for example, but that should not impact the user experience and there should be absolutely no delay in Chrome. I'll file a separate issue for that.

**Comment by muffinresearch** [comment]
See also #3733

**Comment by tofumatt** [comment]
Multiple entry points means loading new JS for each page, right? Numbers obviously win over "Gut feelings" but I imagine that's a small trade-off for first page load to making every other page load slowly. I know our first load is slow (though it's apparently not longer than the old AMO), but every page afterward needs only to make network requests (and JS code loading) for API data and assets (images/icons). That's pretty great, I'd imagine loading new JS per-page would complicate the app and possibly slow down the navigation experience.

How would the new page know how to render loading bars in content until the JS is loaded, for instance? A lot of our navigation is based on having a component's code pre-loaded and ready to render with no data.

**Comment by muffinresearch** [comment]
@tofumatt yep, there's definitely some trade-off concerns. This is why we should absolutely continue the discussions and explore what this would mean before trying it out.

If we come to the conclusion that it adds complication for little real-world benefit then it will be totally fine to not do it at all.

Part of this would be understanding the traffic on different pages. E.g. if there are parts of the site that are seldom visited then maybe those should not be loaded up front, if say because for the most part people don't even look at them. E.g. the About page might be an example of that.

**Comment by muffinresearch** [comment]
Another consideration here is to think about the different types of users. On a desktop with a fast connection the big bundle up front might be better, but for a mobile client it could be a completely different story.

**Comment by willdurand** [comment]
I stumbled upon this [react-router documentation section about code splitting](https://reacttraining.com/react-router/web/guides/code-splitting), but I am not sure this project uses the latest version of that lib.

**Comment by willdurand** [comment]
@muffinresearch I tried to code split a few page components (some error components) in our [`Routes` component](https://github.com/mozilla/addons-frontend/blob/master/src/amo/components/Routes/index.js), just to see the impact on our [Lighthouse](https://developers.google.com/web/tools/lighthouse/) "performance" results. It is easy to drastically reduce the time to first meaningful paint.

Given that we likely need code splitting for the "blog-post style pages", we could maybe start code splitting the error and static pages? We should probably measure the outcome too.

I believe that we could probably rely on preloading (in a second phase) to avoid "loading" delays/flickerings.

**Comment by muffinresearch** [comment]
@willdurand yes that sounds good to me.

**Comment by rebmullin** [comment]
@willdurand, this is something that I am looking at now for the guide pages. Just checking, are you using [the react-router documentation section about code splitting](https://reacttraining.com/react-router/web/guides/code-splitting) for this? 

Should I start with the error & static pages as you mention before setting this up for guide pages?

**Comment by willdurand** [comment]
> @willdurand, this is something that I am looking at now for the guide pages. Just checking, are you using [the react-router documentation section about code splitting](https://reacttraining.com/react-router/web/guides/code-splitting) for this?

I've seen this page. We should try this lib too: https://github.com/smooth-code/loadable-components, because SSR support is a must-have and, for having tried the react-loadable lib, I am not a fan of the server setup. https://github.com/smooth-code/loadable-components seems better.

**Comment by willdurand** [comment]
So, I finally wrote something that works: https://github.com/mozilla/addons-frontend/pull/7017

I use `loadable-components` as linked above. I code splitted only one component and everything seems to work in development and production mode (locally). I added the SRI support and it only works with JS.

**Comment by ioanarusiczki** [comment]
@willdurand I verified the pages on AMO dev (FF63, Chrome, Edge - Win10) as requested by @muffinresearch  [above](https://github.com/mozilla/addons-frontend/issues/3734#issue-270252909).

My test results:
- with FF63 when the links from the content of the pages are clicked and then the user navigates back there's a glitch every time (but not when links from the footer or the FF logo is clicked) This glitch is not present on stage or prod.

![glich - is new](https://user-images.githubusercontent.com/33448286/49583874-25239400-f962-11e8-89af-f5d3d8cc643c.gif)

- with Chrome the glitch is reproducible clicking any link (including those from the footer of the page) and it's more obvious. 

![glich - on chrome](https://user-images.githubusercontent.com/33448286/49584145-01ad1900-f963-11e8-9f1b-7f6379a27c09.gif)

- the FF and Chrome issue I mentioned is intermittent with Edge (it works quite well).
- the glitch is also reproducible on Android (I verified it with FF and Chrome and it's more obvious with Chrome) 
- I did not have delay problems.

**Comment by kumar303** [comment]
@ioanarusiczki thanks for catching these issues. The glitch is a minor issue so we don't need to back out this patch. Because we're not backing out the patch, I'll close the issue. Please open a new issue so we can investigate the glitch.
